File: OpenWA/sdk/python/openwa/resources/contacts.py
"""Contacts resource — contact lookup and management.

Backed by ``src/modules/contact/contact.controller.ts``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, TypedDict

from .._http import quote_segment
from ..types import (
    UpsertContactRequest,
    CheckNumberResponse,
    ContactPhoneResponse,
    ContactRecord,
    ProfilePictureResponse,
    ProfilePicturesResponse,
    SuccessResult,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class ListContactsQuery(TypedDict, total=False):
    limit: int
    offset: int


class ContactsResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str, query: ListContactsQuery | None = None) -> list[ContactRecord]:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/contacts", query=query)

    def get(self, session_id: str, contact_id: str) -> ContactRecord:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}")

    def check(self, session_id: str, number: str) -> CheckNumberResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/contacts/check/{quote_segment(number)}")

    def profile_picture(self, session_id: str, contact_id: str) -> ProfilePictureResponse:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}/profile-picture"
        )

    def profile_pictures(self, session_id: str, ids: list[str]) -> ProfilePicturesResponse:
        """Batch-resolve profile picture URLs for up to 50 contacts in one request.
        Returns a map of contact id → URL (None when a lookup fails)."""
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/contacts/profile-pictures", query={"ids": ",".join(ids)}
        )

    def phone(self, session_id: str, contact_id: str) -> ContactPhoneResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}/phone")

    def block(self, session_id: str, contact_id: str) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}/block")

    def upsert(self, session_id: str, contact_id: str, body: UpsertContactRequest) -> SuccessResult:
        """Save a contact to the addressbook, or edit an existing entry (OPERATOR)."""
        return self._http.request(
            "PUT", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}", body=body
        )

    def delete(self, session_id: str, contact_id: str) -> SuccessResult:
        """Remove a contact from the addressbook (OPERATOR)."""
        return self._http.request(
            "DELETE", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}"
        )

    def unblock(self, session_id: str, contact_id: str) -> SuccessResult:
        return self._http.request("DELETE", f"/api/sessions/{quote_segment(session_id)}/contacts/{quote_segment(contact_id)}/block")

File: OpenWA/sdk/javascript/src/resources/contacts.ts
/**
 * Contacts resource — contact lookup and management.
 *
 * Backed by `src/modules/contact/contact.controller.ts`.
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type {
  UpsertContactRequest,
  CheckNumberResponse,
  ContactPhoneResponse,
  ContactRecord,
  ProfilePictureResponse,
  ProfilePicturesResponse,
  SuccessResult,
} from '../types.js';

export interface ListContactsQuery {
  limit?: number;
  offset?: number;
}

export class ContactsResource {
  constructor(private readonly client: OpenWAClient) {}

  /** List contacts known to the session. */
  list(sessionId: string, query?: ListContactsQuery): Promise<ContactRecord[]> {
    return this.client.request<ContactRecord[]>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts`,
      query,
    });
  }

  /** Get details for a single contact by id (JID). */
  get(sessionId: string, contactId: string): Promise<ContactRecord> {
    return this.client.request<ContactRecord>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}`,
    });
  }

  /** Check whether a phone number is registered on WhatsApp. */
  check(sessionId: string, number: string): Promise<CheckNumberResponse> {
    return this.client.request<CheckNumberResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/check/${encodeSegment(number)}`,
    });
  }

  /** Get the contact's profile picture URL (or null). */
  profilePicture(sessionId: string, contactId: string): Promise<ProfilePictureResponse> {
    return this.client.request<ProfilePictureResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}/profile-picture`,
    });
  }

  /**
   * Batch-resolve profile picture URLs for up to 50 contacts in one request.
   * Returns a map of contact id → URL (null when a lookup fails).
   */
  profilePictures(sessionId: string, ids: string[]): Promise<ProfilePicturesResponse> {
    return this.client.request<ProfilePicturesResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/profile-pictures`,
      query: { ids: ids.join(',') },
    });
  }

  /** Resolve a contact id (e.g. a `@lid`) to a phone number. */
  phone(sessionId: string, contactId: string): Promise<ContactPhoneResponse> {
    return this.client.request<ContactPhoneResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}/phone`,
    });
  }

  /** Block a contact. Requires an OPERATOR-level key. */
  block(sessionId: string, contactId: string): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}/block`,
    });
  }

  /**
   * Save a contact to the account's addressbook, or edit an existing entry.
   * Requires an OPERATOR-level key.
   */
  upsert(sessionId: string, contactId: string, body: UpsertContactRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'PUT',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}`,
      body,
    });
  }

  /** Remove a contact from the account's addressbook. Requires an OPERATOR-level key. */
  delete(sessionId: string, contactId: string): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'DELETE',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}`,
    });
  }

  /** Unblock a contact. Requires an OPERATOR-level key. */
  unblock(sessionId: string, contactId: string): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'DELETE',
      path: `/api/sessions/${encodeSegment(sessionId)}/contacts/${encodeSegment(contactId)}/block`,
    });
  }
}
